### DISCUSSION TITLE
`Framework preset: Other` does not work properly when having `next` dependency as well

### DISCUSSION THREAD

**Discussion by lambrospetrou** [vercel/vercel#4132] [General] [answered]
My project uses NextJS but I am using `next export` in order to have a fully functional static website, instead of the default setup.

I setup the `Build & Development Settings` settings to have my custom build command (`make`), the directory of the static output (`_site`), and the `Framework preset` to `Other`.

From the logs I see that the build completes execution fine, but at the end it seems that the deployment *still* tries to treat the project as a normal NextJS project and tries to load the `routes-manifest.json` file, which does not exist since I am using the static export.

The error seems to be coming from [@now/next](https://github.com/zeit/now/blob/57d25b184b63d262240856e40646ba9e1e4125e2/packages/now-next/src/utils.ts#L343).

I also tried setting the `Framework preset` to `Hugo` and couple of others but it seems that when the `package.json` contains `next` it always runs the `@now/next` builder/plugin which causes my deployments to fail.

Did anyone encounter this behavior? It seems to be buggy behavior from the build script since I thought specifying `Other` will give full control to the provided build command?

It would be nice to avoid doing hacky workarounds...

Thanks

**Comment by styfle** [comment]
This is similar to the discussion in #4138

In your case, I think you probably should select Next.js from the Framework Preset dropdown and leave the Output Directory as is (which is `public`). Then you can override the Build Command as long as you have `next build` or `next export` in there somewhere.

**Comment by lambrospetrou** [comment]
Hi,

Yes I can find a workaround to make my current situation working, but this
is 100% a bug in the builders, because it might be working here but with a
different more complex flow it wouldn't.

Is there a way to open a proper ticket on the Github repo so that more
people get visibility or at least know the 'Other' is not working properly?

I wasted a few hours before finding a solution on my own before opening a case with Vercel.

The output is `out` not `public` BTW.

Thanks for your help

**Comment by styfle** [comment]
Thanks for the feedback!

We updated the error message in PR #4161.

**Comment by lambrospetrou** [comment]
@styfle Thanks for the more clarified error message but similar to the discussion you linked above, it's completely missing the reported issue.

I will try to restate the issue in case I didn't explain it properly initially. However, if my understanding of the `Other` preset is completely wrong please let me know, but according to the [Build step docs](https://vercel.com/docs/v2/build-step#build-command) my assumption about the expected behavior seems to be correct.

Based on the docs, selecting `Other` as the preset framework means it's not any of the predefined ones, and therefore I will have to specify the custom command to use to build the project and optionally override the output directory if it's not `public`.

From this I understand and expect that selecting `Other` will allow me to use any series of commands to generate the output directory of my choice regardless of the libraries I use in my `package.json`. In my case this command is literally just `make`, and the `makefile` I have calls some NPM targets I have defined. For example, in my project `next` is just a part of the build toolchain, since I also do some post-processing after `next build && npm export` in order to add additional files into the output folder or even modify some of them. I would expect that it has no impact no matter what tools/scripts/libraries are being called in the `Build command` as long as at the end they generate the directory as specified in `Output directory`.

Then once the build command finishes successfully, I expect that Vercel will take the output directory and distribute it to the edge locations and serve it as a normal static website.

So, the issue is that even though I provided **override values** for both, `Build command`, and `Output directory`, and even though I also specified the `Other` preset, Vercel still tries to be smart and considers the project to be a standard `next` project and uses the `@now/next` builder to process the build commands, just because it found it in the `package.json`.

What if I have both `next` and `gatsby` and any other framework in my `package.json`. I do understand that it's not common, and it could even be a bad thing, but this does not make the current behavior correct :) There are valid use-cases to have some post-processing after calling `next`, using custom output directories, etc.

If I completely missed the point of the `Other` option and the override values please correct me. The `Other` option as I expected it to work satisfies all my needs for my upcoming projects but if it always assumes it's a `next` project it means I will need to do some **hacky workarounds just to play nice.**

Thanks a lot for any help!

**Reply by dbauszus-glx** [reply]
@styfle  Is it possible to set the framework preset in the CLI/vercel.json or only through the dashboard?

**Reply by styfle** [reply]
@dbauszus-glx Yes, you can use the `vercel link` command to create a new project which will ask for the framework preset.

**Reply by dbauszus-glx** [reply]
@styfle Unfortunately `vercel link` fails straight off the bat. Is there any documentation for that option? A google for `vercel link` returns nothing.

```
Error! request to https://api.vercel.com/v1/teams? failed, reason: self signed certificate
```

**Reply by dbauszus-glx** [reply]
@styfle To clarify... This is for a new deployment. When I first deploy I want to set the framework, and not rely on the auto-detect which isn't working. Ideally I just would like to not have auto-detect and just have other as default. Or if at least I could overwrite the build and dev settings. But that option is also broken in the CLI. https://github.com/vercel/vercel/discussions/5162

**Comment by styfle** [comment]
Ok I see what you mean. Auto-detecting Next.js is unexpected when manually choosing "Other" as the framework preset.

I think the root cause is that the framework preset is only used by `@now/static-build` but Next.js is special cased to use `@now/next`. I'll create a ticket for this bug and we'll get it fixed, thanks!

**Reply by lambrospetrou** [reply]
Perfect, I am glad this is not how it should work :) (just for the sake of my understanding that is)

**Comment by lambrospetrou** [comment]
@styfle Is there any progress on this issue? Any ticket/issue I can keep track?

Thanks

**Reply by styfle** [reply]
No progress yet. Our tickets are tracked internally. This one https://vercel.atlassian.net/browse/PRODUCT-2637

**Comment by styfle** [comment, marked as answer]
This issue has been fixed in PR #5009, thanks!

**Reply by lambrospetrou** [reply]
Awesome, I just tried it and it worked! Thanks
